### PR TITLE
perf(ruleeng): employ `emqx_topic_index` to speed up topic matching

### PR DESCRIPTION
**Pull request by keynslug**
Made an attempt to speed up the rule engine hot-path in spare time. The meat is in `emqx_topic_index` which utilizes ordered set semantics to match topics to topic filters. The `emqx_topic_index_SUITE` is basically just a `emqx_trie_SUITE` retrofitted to the new API, to verify they both give the same results with same testing data.

### <samp>🤖 Generated by Copilot at 3ba3fb4</samp>

This pull request introduces a new module `emqx_topic_index` that provides a fast and accurate topic index for matching topics to topic filters. It also refactors the `emqx_rule_engine` module to use the topic index for rule lookup and updates the topic index when rules are added or removed. Additionally, it defines a macro for the topic index table name and adds a test suite for the new module.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] ~~If there should be document changes, a PR to emqx-docs.git is sent, or a jira ticket is created to follow up~~
- [x] Schema changes are backward compatible

### CODE REVIEW

**Comment by keynslug** [comment]
> Couldn't we convert the filter to a matchspec and make select?

Just to clarify: we need to tell if a topic matches some subset of topic filters stored in a ETS table, not vice versa. So we would need to convert topic `a/b/c` to something like `[["a", "b", "c"], ["a", "b", "c", "#"], ["a", "b", "+"], ["a", "b", "+", "#"], ...]` which explodes quite quickly. Maybe I'm missing something though.

**Comment by savonarola** [comment]
Ah, I see; we are searching filter by topic, not vice versa.

File: apps/emqx/src/emqx_topic_index.erl [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+-export([get_id/1]).
+-export([get_topic/1]).
+-export([get_record/2]).
+
+-type key(ID) :: [binary() | '+' | '#' | {ID}].
+-type match(ID) :: key(ID).
+
+new() ->
+    ets:new(?MODULE, [public, ordered_set, {write_concurrency, true}]).
```
Comment (JimMoen) [inline apps/emqx/src/emqx_topic_index.erl:48]: `new/0` is not used anywhere except in test cases.

File: apps/emqx/src/emqx_topic_index.erl [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    stop.
+
+match_init(Topic) ->
+    case emqx_topic:words(Topic) of
+        [W = <<"$", _/bytes>> | Rest] ->
+            % NOTE
+            % This will effectively skip attempts to match special topics to `#` or `+/...`.
+            {Rest, [W]};
+        Words ->
+            {Words, []}
```
Comment (JimMoen) [inline apps/emqx/src/emqx_topic_index.erl:136]:
```suggestion
        [W | RestWords] ->
            {RestWords, [W]}
```

**Review by JimMoen — CHANGES_REQUESTED** [review]

**Comment by JimMoen** [comment]
And Plz change base to `release-51`. We'll merge it into next release.

**Comment by JimMoen** [comment]
Cherry picked and pushed on #11282

**Pull request closed without merging**